File: controllers/supplier_controller.py
from typing import Dict, List, Any
from utils.database import database
from utils.logger_utils import log_error, log_info
from sqlalchemy.exc import IntegrityError
from datetime import datetime
from utils.meilisearch import client
from schemas.supplier_schema import SupplierCreate, SupplierUpdate, SupplierSearchDocument
from repository.supplier_repository import SupplierRepository

class SupplierController:
    @staticmethod
    async def create_supplier(supplier_data: dict, user_id: int) -> Dict[str, Any]:
        """
        Create a new supplier in the database.
        """
        log_info(f"Creating supplier with data: {supplier_data}")
        try:
            # Add user ID to supplier data
            supplier_data["createdBy"] = user_id
            supplier_data['isDelete'] = False
            
            # Validate and create supplier model
            supplier_create = SupplierCreate(**supplier_data)
            
            # Use repository to create supplier
            result = await SupplierRepository.create(supplier_create)
            
            if "error" in result:
                log_error(f"Error creating supplier: {result['error']}")
                return {"error": result["error"], "status": result["status"]}
            
            supplier_id = result["supplier_id"]
            
            # Index in Meilisearch
            await SupplierController._index_supplier_in_search(supplier_id, supplier_data)
            
            log_info(f"Supplier created successfully with ID: {supplier_id}")
            return result
            
        except Exception as e:
            log_error(f"Unexpected error: {str(e)}")
            return {"error": "Internal server error.", "status": 500}

    @staticmethod
    async def get_supplier(supplier_id: int) -> Dict[str, Any]:
        """
        Get a supplier by ID from the database.
        """
        log_info(f"Fetching supplier with ID: {supplier_id}")
        try:
            supplier = await SupplierRepository.get_by_id(supplier_id)
            
            if not supplier:
                return {"error": "Supplier not found.", "status": 404}
                
            return supplier.model_dump()
            
        except Exception as e:
            log_error(f"Error fetching supplier: {str(e)}")
            return {"error": "Internal server error.", "status": 500}

    @staticmethod
    async def get_suppliers(keyword: str = None, page: int = 1, pageSize: int = 10) -> Dict[str, Any]:
        """
        Get suppliers from the database with Meilisearch fallback.
        """
        log_info(f"Fetching suppliers with keyword: {keyword}, page: {page}, page_size: {pageSize}")
        try:
            # Try Meilisearch first
            try:
                result = client.index("suppliers").search(
                    keyword, 
                    {"limit": pageSize, "offset": (page - 1) * pageSize}
                )

                if result["hits"]:
                    return {
                        "data": result["hits"],
                        "count": result["estimatedTotalHits"],
                        "page": page,
                        "page_size": pageSize
                    }
            except Exception as search_error:
                log_error(f"Meilisearch error, falling back to database: {str(search_error)}")
            
        except Exception as e:
            log_error(f"Error fetching suppliers: {str(e)}")
            return {"error": "Internal server error.", "status": 500}

    @staticmethod
    async def update_supplier(supplier_data: dict, user_id: int) -> Dict[str, Any]:
        """
        Update an existing supplier in the database.
        """
        log_info(f"Updating supplier with data: {supplier_data}")
        try:
            if "id" not in supplier_data:
                return {"error": "Supplier ID is required for update", "status": 400}
                
            supplier_id = supplier_data["id"]
            
            # Check if supplier exists
            existing_supplier = await SupplierRepository.get_by_id(supplier_id)
            if not existing_supplier:
                return {"error": "Supplier not found", "status": 404}
            
            # Add updatedBy user ID
            supplier_data["updatedBy"] = user_id
            
            # Validate update data
            supplier_update = SupplierUpdate(**supplier_data)
            
            # Update supplier
            result = await SupplierRepository.update(supplier_id, supplier_update)
            
            if "error" in result:
                log_error(f"Error updating supplier: {result['error']}")
                return {"error": result["error"], "status": result["status"]}
            
            # Update search index
            await SupplierController._index_supplier_in_search(supplier_id, supplier_data)
            
            log_info(f"Supplier updated successfully with ID: {supplier_id}")
            return result
            
        except IntegrityError as e:
            log_error(f"Integrity error while updating supplier: {str(e)}")
            return {"error": "Supplier already exists.", "status": 400}
        except Exception as e:
            log_error(f"Unexpected error while updating supplier: {str(e)}")
            return {"error": "Internal server error.", "status": 500}

    @staticmethod
    async def delete_supplier(supplier_id: int, user_id: int) -> Dict[str, Any]:
        """
        Soft delete a supplier.
        """
        log_info(f"Deleting supplier with ID: {supplier_id}")
        try:
            # Check if supplier exists
            existing_supplier = await SupplierRepository.get_by_id(supplier_id)
            if not existing_supplier:
                return {"error": "Supplier not found", "status": 404}
            
            # Soft delete supplier
            result = await SupplierRepository.soft_delete(supplier_id, user_id)
            
            if "error" in result:
                return {"error": result["error"], "status": result["status"]}
            
            # Remove from search index
            try:
                client.index("suppliers").delete_document(supplier_id)
            except Exception as e:
                log_error(f"Error removing supplier from search index: {str(e)}")
            
            log_info(f"Supplier deleted successfully with ID: {supplier_id}")
            return result
            
        except Exception as e:
            log_error(f"Error deleting supplier: {str(e)}")
            return {"error": "Internal server error.", "status": 500}

    @staticmethod
    async def _index_supplier_in_search(supplier_id: int, supplier_data: dict):
        """
        Index supplier in Meilisearch.
        """
        try:
            search_doc = SupplierSearchDocument(
                id=supplier_id,
                name=f"{supplier_data['name']}, {supplier_data.get('prefix', '')}",
                address=supplier_data["address"],
                city=supplier_data["city"],
                province=supplier_data["province"],
                phoneNumber=supplier_data["phoneNumber"],
                email=supplier_data.get("email"),
                npwp=supplier_data.get("npwp"),
                itemsSold=supplier_data["itemsSold"].split(",") if supplier_data["itemsSold"] else [],
                serviceArea=supplier_data["serviceArea"].split(",") if supplier_data["serviceArea"] else []
            )
            
            client.index("suppliers").add_documents([search_doc.model_dump()])
            
        except Exception as e:
            log_error(f"Error indexing supplier in search: {str(e)}")

File: src/controllers/supplier.controller.ts
import { SupplierRepository } from "../repository/supplier"
import { UserRepository } from "../repository/user"
import { meili } from "../utils/meilisearch"
import { logError, logInfo } from "../utils/logger"
import { paginationMeta } from "../utils/pagination"
import { SupplierModel } from "../models/supplier.model"
import { UserModel } from "../models/user.model"
import type { ISupplierCreate, ISupplierOutput, ISupplierUpdate, ISupplierView } from "../interfaces/supplier.interfaces"
import type { IUserView } from "../interfaces/user.interfaces"

const index = meili.index("suppliers")

async function indexSupplier(supplier: Record<string, unknown>) {
  try {
    const doc = {
      ...supplier,
      phone_number: supplier.phoneNumber,
      name: `${supplier.name}, ${supplier.prefix ?? ""}`,
      items_sold: supplier.itemsSold ? String(supplier.itemsSold).split(",").map((s) => s.trim()) : [],
      service_area: supplier.serviceArea ? String(supplier.serviceArea).split(",").map((s) => s.trim()) : [],
      is_blacklisted: supplier.isBlackListed ?? false,
    }
    for (const [k, v] of Object.entries(doc)) {
      if (v instanceof Date) (doc as Record<string, unknown>)[k] = v.toISOString()
      else if (v === null) (doc as Record<string, unknown>)[k] = ""
    }
    await index.addDocuments([doc])
  } catch (e) {
    logError(`Meilisearch indexing error: ${e}`)
  }
}

async function buildUserView(userId: number | null | undefined): Promise<IUserView> {
  if (!userId) return UserModel.toView(null)
  const user = await UserRepository.getById(userId)
  return UserModel.toView(user)
}

export class SupplierController {
  static async create(data: ISupplierCreate, userId: number): Promise<{ kind: 'success'; data: ISupplierOutput } | { kind: 'error'; error: string; status: number }> {
    try {
      const payload = SupplierModel.createPayload(data)
      const supplier = await SupplierRepository.create({ ...payload, createdBy: userId, createdAt: new Date() })
      void indexSupplier(supplier as Record<string, unknown>)
      const userView = await buildUserView(userId)
      logInfo(`Supplier created: ${supplier.id}`)
      return { kind: 'success', data: SupplierModel.toOutput(supplier, userView, userView) }
    } catch (e) {
      logError(`Error creating supplier: ${e}`)
      return { kind: 'error', error: "Failed to create supplier", status: 500 }
    }
  }

  static async getAll(page: number, pageSize: number, keyword?: string, sortBy?: string, sortDir?: string): Promise<{ data: ISupplierView[]; meta: any } | { error: string; status: number }> {
    if (page < 1) return { error: "Page must be >= 1", status: 400 }
    try {
      const { data, total } = await SupplierRepository.getAll(page, pageSize, keyword, sortBy, sortDir)
      const transformedData = data.map(SupplierModel.toView)
      return { data: transformedData, meta: paginationMeta(total, page, pageSize) }
    } catch (e) {
      logError(`Error fetching suppliers: ${e}`)
      return { error: "Failed to fetch suppliers", status: 500 }
    }
  }

  static async getById(id: number): Promise<{ kind: 'success'; data: ISupplierOutput } | { error: string; status: number }> {
    try {
      const supplier = await SupplierRepository.getById(id)
      if (!supplier) return { error: "Supplier not found", status: 404 }
      const [createdBy, updatedBy] = await Promise.all([
        buildUserView(supplier.createdBy),
        buildUserView(supplier.updatedBy),
      ])
      return { kind: 'success', data: SupplierModel.toOutput(supplier, createdBy, updatedBy) }
    } catch (e) {
      logError(`Error fetching supplier ${id}: ${e}`)
      return { error: "Failed to fetch supplier", status: 500 }
    }
  }

  static async update(id: number, data: ISupplierUpdate, userId: number): Promise<{ kind: 'success'; data: ISupplierOutput } | { error: string; status: number }> {
    try {
      const existing = await SupplierRepository.getById(id)
      if (!existing) return { error: "Supplier not found", status: 404 }

      const payload = SupplierModel.updatePayload(data)
      const updated = await SupplierRepository.update(id, { ...payload, updatedBy: userId, updatedAt: new Date() })
      void indexSupplier(updated as Record<string, unknown>)
      const [createdBy, updatedBy] = await Promise.all([
        buildUserView(existing.createdBy),
        buildUserView(userId),
      ])
      logInfo(`Supplier updated: ${id}`)
      return { kind: 'success', data: SupplierModel.toOutput(updated, createdBy, updatedBy) }
    } catch (e) {
      logError(`Error updating supplier ${id}: ${e}`)
      return { error: "Failed to update supplier", status: 500 }
    }
  }

  static async setBlackList(id: number, isBlackListed: boolean, userId: number) {
    try {
      const existing = await SupplierRepository.getById(id)
      if (!existing) return { error: "Supplier not found", status: 404 }

      const updated = await SupplierRepository.update(id, { isBlackListed, updatedBy: userId, updatedAt: new Date() })
      void indexSupplier(updated as Record<string, unknown>)
      logInfo(`Supplier ${id} blacklist set to ${isBlackListed}`)
      return { message: "Supplier blacklist updated" }
    } catch (e) {
      logError(`Error setting blacklist for supplier ${id}: ${e}`)
      return { error: "Failed to update blacklist", status: 500 }
    }
  }

  static async delete(id: number, userId: number) {
    try {
      const existing = await SupplierRepository.getById(id)
      if (!existing) return { error: "Supplier not found", status: 404 }

      await SupplierRepository.softDelete(id, userId)
      await index.deleteDocument(id)
      logInfo(`Supplier deleted: ${id}`)
      return { message: "Supplier deleted successfully" }
    } catch (e) {
      logError(`Error deleting supplier ${id}: ${e}`)
      return { error: "Failed to delete supplier", status: 500 }
    }
  }
}
